File: db4on/core/src/core/com/db4o/cs/messages/MsgObject.cs
namespace com.db4o.cs.messages
{
	public class MsgObject : com.db4o.cs.messages.MsgD
	{
		private const int LENGTH_FOR_ALL = com.db4o.YapConst.ID_LENGTH + (com.db4o.YapConst
			.INT_LENGTH * 3);

		private const int LENGTH_FOR_FIRST = LENGTH_FOR_ALL;

		internal int _id;

		internal int _address;

		internal virtual com.db4o.cs.messages.MsgD GetWriter(com.db4o.YapWriter bytes, int[]
			 prependInts)
		{
			int lengthNeeded = bytes.GetLength() + LENGTH_FOR_FIRST;
			if (prependInts != null)
			{
				lengthNeeded += (prependInts.Length * com.db4o.YapConst.INT_LENGTH);
			}
			int embeddedCount = bytes.EmbeddedCount();
			if (embeddedCount > 0)
			{
				lengthNeeded += (LENGTH_FOR_ALL * embeddedCount) + bytes.EmbeddedLength();
			}
			com.db4o.cs.messages.MsgD message = GetWriterForLength(bytes.GetTransaction(), lengthNeeded
				);
			if (prependInts != null)
			{
				for (int i = 0; i < prependInts.Length; i++)
				{
					message._payLoad.WriteInt(prependInts[i]);
				}
			}
			message._payLoad.WriteInt(embeddedCount);
			bytes.AppendTo(message._payLoad, -1);
			return message;
		}

		public override com.db4o.cs.messages.MsgD GetWriter(com.db4o.YapWriter bytes)
		{
			return GetWriter(bytes, null);
		}

		public virtual com.db4o.cs.messages.MsgD GetWriter(com.db4o.YapClass a_yapClass, 
			com.db4o.YapWriter bytes)
		{
			if (a_yapClass == null)
			{
				return GetWriter(bytes, new int[] { 0 });
			}
			return GetWriter(bytes, new int[] { a_yapClass.GetID() });
		}

		public virtual com.db4o.cs.messages.MsgD GetWriter(com.db4o.YapClass a_yapClass, 
			int a_param, com.db4o.YapWriter bytes)
		{
			return GetWriter(bytes, new int[] { a_yapClass.GetID(), a_param });
		}

		public com.db4o.YapWriter Unmarshall()
		{
			return Unmarshall(0);
		}

		public com.db4o.YapWriter Unmarshall(int addLengthBeforeFirst)
		{
			_payLoad.SetTransaction(GetTransaction());
			int embeddedCount = _payLoad.ReadInt();
			int length = _payLoad.ReadInt();
			if (length == 0)
			{
				return null;
			}
			_id = _payLoad.ReadInt();
			_address = _payLoad.ReadInt();
			if (embeddedCount == 0)
			{
				_payLoad.RemoveFirstBytes(LENGTH_FOR_FIRST + addLengthBeforeFirst);
			}
			else
			{
				_payLoad._offset += length;
				com.db4o.YapWriter[] embedded = new com.db4o.YapWriter[embeddedCount + 1];
				embedded[0] = _payLoad;
				new com.db4o.YapWriter(_payLoad, embedded, 1);
				_payLoad.Trim4(LENGTH_FOR_FIRST + addLengthBeforeFirst, length);
			}
			_payLoad.UseSlot(_id, _address, length);
			return _payLoad;
		}

		public virtual void PayLoad(com.db4o.YapWriter writer)
		{
			_payLoad = writer;
		}

		public virtual int GetId()
		{
			return _id;
		}
	}
}

File: db4oj/core/src/com/db4o/internal/cs/messages/MsgObject.java
/* Copyright (C) 2004   db4objects Inc.   http://www.db4o.com */

package com.db4o.internal.cs.messages;

import com.db4o.internal.*;
import com.db4o.internal.slots.*;


public class MsgObject extends MsgD {
	
	private static final int LENGTH_FOR_ALL = Const4.ID_LENGTH + (Const4.INT_LENGTH * 2);
	private static final int LENGTH_FOR_FIRST = LENGTH_FOR_ALL;
	
	private int _id;
	private int _address;
	
	final MsgD getWriter(Pointer4 pointer, StatefulBuffer bytes,int[] prependInts) {
		int lengthNeeded = bytes.length() + LENGTH_FOR_FIRST;
		if(prependInts != null){
			lengthNeeded += (prependInts.length * Const4.INT_LENGTH);
		}
		MsgD message = getWriterForLength(bytes.getTransaction(), lengthNeeded);
		if(prependInts != null){
		    for (int i = 0; i < prependInts.length; i++) {
		        message._payLoad.writeInt(prependInts[i]);    
            }
		}
		message._payLoad.append(pointer, bytes);
		return message;
	}

	final public MsgD getWriter(StatefulBuffer buffer) {
		return getWriter(buffer.pointer(), buffer, null);
	}
	
	public final MsgD getWriter(Pointer4 pointer, ClassMetadata a_yapClass, StatefulBuffer bytes) {
        if(a_yapClass == null){
            return getWriter(pointer, bytes, new int[]{0});
        }
		return getWriter(pointer, bytes, new int[]{ a_yapClass.getID()});
	}
	
	public final MsgD getWriter(Pointer4 pointer, ClassMetadata classMetadata, int param, StatefulBuffer buffer) {
		return getWriter(pointer, buffer, new int[]{ classMetadata.getID(), param});
	}
	
	public final StatefulBuffer unmarshall() {
		return unmarshall(0);
	}

	public final StatefulBuffer unmarshall(int addLengthBeforeFirst) {
		_payLoad.setTransaction(transaction());
		
		int length = _payLoad.readInt();
		if (length == 0) {
			return null;  // does this happen ?
		}
		_id = _payLoad.readInt();
		_address = _payLoad.readInt();
		_payLoad.removeFirstBytes(LENGTH_FOR_FIRST + addLengthBeforeFirst);
		_payLoad.useSlot(_id, _address, length);
		return _payLoad;
	}

	public int getId() {
		return _id;
	}
}
